extensions/amp-image-viewer/0.1/amp-image-viewer.js

```diff
@@ -119,61 +109,85 @@ export class ImageViewer {
     this.maxX_ = 0;
     /** @private {number} */
     this.maxY_ = 0;
-    /** @private {?./gesture.Gestures} */
+
+    /** @private {?../../../src/gesture.Gestures} */
     this.gestures_ = null;
-    /** @private {?function()} */
-    this.unlistenOnSwipePan_ = null;
 
-    /** @private {?./motion.Motion} */
+    /** @private {?../../../src/motion.Motion} */
     this.motion_ = null;
 
-    this.setupGestures_();
+    /** @private {?Element} */
+    this.sourceElement_ = null;
   }
 
-  /**
-   * Returns the root element of the image viewer.
-   * @return {!Element}
-   */
-  getElement() {
-    return this.viewer_;
+  /** @override */
+  buildCallback() {
+    this.vsync_ = this.getVsync();
+    this.element.classList.add('i-amphtml-image-lightbox-viewer');
+    const children = this.getRealChildren();
+    dev().assert(children.length == 1);
+    this.sourceElement_ = children[0];
+    this.setAsOwner(this.sourceElement_);
   }
 
-  /**
-   * Returns the img element of the image viewer.
-   * @return {!Element}
-   */
-  getImage() {
-    return this.image_;
+  /** @override */
+  layoutCallback() {
+    let loadPromise = Promise.resolve();
+    if (this.sourceElement_) {
+      this.scheduleLayout(this.sourceElement_);
+      loadPromise = this.sourceElement_.signals()
+          .whenSignal(CommonSignals.LOAD_END);
+    }
+    return loadPromise
+        .then(() => {
+          return this.vsync_.mutatePromise(() => {
+            if (!this.image_) {
+              this.image_ = this.win.document.createElement('img');
+              this.image_.classList.add('i-amphtml-image-viewer-image');
+
+              this.init_();
+              this.element.appendChild(this.image_);
+
+              this.element.removeChild(this.sourceElement_);
+              this.sourceElement_ = null;
+            }
+            this.setupGestures_();
+            this.measure();
+            this.updateSrc_();
+            this.registerOnResizeHandler_();
+          });
+        });
   }
 
-  /**
-   * Returns the boundaries of the viewer.
-   * @return {!./layout-rect.LayoutRectDef}
-   */
-  getViewerBox() {
-    return this.viewerBox_;
+  /** @override */
+  unlayoutCallback() {
+    this.gestures_.cleanup();
+    this.cleanupOnResizeHandler_();
```

Comment:
Maybe set `this.image_ = null` to free up the memory. It should be safe since you reassign it again in `layoutCallback`

Reply:
Is it worth doing this? So I incase the creation and assignment of `this.image_` in an if block because we only ever want to do it once. If we garbage collected this, we would have to create, append, remove every time `this.image_` goes through a layout/relayout cycle. I'm actually not very sure about the memory / performance tradeoffs here at all. Do you know what best practices are here?